booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java

```diff
@@ -0,0 +1,645 @@
+    "Validate allowed shipment cutoff codes",
+    (mav) -> mav.submitAllMatching("shipmentCutOffTimes.*.cutOffDateTimeCode"),
+    JsonAttribute.matchedMustBeDatasetKeywordIfPresent(BookingDataSets.CUTOFF_DATE_TIME_CODES)
+  );
+
+  private static final JsonContentCheck VALIDATE_SHIPMENT_CUTOFF_TIME_CODE = JsonAttribute.customValidator(
+    "validate shipment cutOff Date time code",
+    (body) -> {
+      var shipmentCutOffTimes = body.path("shipmentCutOffTimes");
+      var receiptTypeAtOrigin = body.path("receiptTypeAtOrigin").asText("");
+      var issues = new LinkedHashSet<String>();
+      var cutOffDateTimeCodes = StreamSupport.stream(shipmentCutOffTimes.spliterator(), false)
+        .map(p -> p.path("cutOffDateTimeCode"))
+        .filter(JsonNode::isTextual)
+        .map(n -> n.asText(""))
+        .collect(Collectors.toSet());
+      if (receiptTypeAtOrigin.equals("CFS") && !cutOffDateTimeCodes.contains("LCL")) {
+        issues.add("cutOffDateTimeCode 'LCL' must be present when receiptTypeAtOrigin is CFS");
+      }
+      return issues;
+    }
+  );
+
+  private static final Consumer<MultiAttributeValidator> ALL_AMF = (mav) -> mav.submitAllMatching("advanceManifestFilings.*");
+
+  private static final JsonContentCheck ADVANCED_MANIFEST_FILING_CODES_UNIQUE = JsonAttribute.allIndividualMatchesMustBeValid(
+    "The combination of 'countryCode' and 'manifestTypeCode' in 'advanceManifestFilings' must be unique",
+    ALL_AMF,
+    JsonAttribute.unique("countryCode", "manifestTypeCode")
+  );
+  private static final Consumer<MultiAttributeValidator> ALL_SHIPMENT_CUTOFF_TIMES = (mav) -> mav.submitAllMatching("shipmentCutOffTimes.*");
+
+  private static final JsonContentCheck SHIPMENT_CUTOFF_TIMES_UNIQUE = JsonAttribute.allIndividualMatchesMustBeValid(
+    "in 'shipmentCutOfftimes' cutOff date time code must be unique",
+    ALL_SHIPMENT_CUTOFF_TIMES,
+    JsonAttribute.unique("cutOffDateTimeCode")
+  );
+  private static final JsonContentCheck AMF_CC_MTC_COMBINATION_VALIDATIONS = JsonAttribute.allIndividualMatchesMustBeValid(
+    "Validate combination of 'countryCode' and 'manifestTypeCode' in 'advanceManifestFilings'",
+    ALL_AMF,
+    JsonAttribute.combineAndValidateAgainstDataset(BookingDataSets.AMF_CC_MTC_COMBINATIONS, "countryCode", "manifestTypeCode")
+  );
+
+  private static final JsonContentCheck VALIDATE_SHIPMENT_LOCATIONS = JsonAttribute.customValidator(
+    "Validate shipmentLocations",
+    body -> {
+      var issues = new LinkedHashSet<String>();
+      var shipmentLocations = body.path("shipmentLocations");
+      var receiptTypeAtOrigin = body.path("receiptTypeAtOrigin").asText("");
+      var polNode =
+        StreamSupport.stream(shipmentLocations.spliterator(), false)
+          .filter(o ->  o.path("locationTypeCode").asText("").equals("POL")
+            || o.path("locationTypeCode").asText("").equals("PRE") )
+          .findFirst()
+          .orElse(null);
```

Comment:
This does not always fetch a POL. If both PRE and POL are present in the list - it is undefined which of them is returned in the polNode...?

Reply:
The shiplocations element can either be PRE or POL for PoL Node. I didnt get your question

Reply:
My suggestion was to not call it `polNode` as it is sometimes a `PRE` - but never mind - the code works and your error message in line 321 is correct :-)